[ja]Updated /content/ja/docs/concepts/services-networking/ingress.md file (#37075)
kubernetes/website: content/ja/docs/concepts/services-networking/ingress.md

{{< codenew file="service/networking/minimal-ingress.yaml" >}}

> Ingressには`apiVersion`、`kind`,
> `metadata`や`spec`フィールドが必要です。Ingressオブジェクトの名前は、有効な[DNSサブドメイン名](/ja/docs/concepts/overview/working-with-objects/names#dns-subdomain-names)である必要があります。設定ファイルに関する一般的な情報は、[アプリケーションのデプロイ](/ja/docs/tasks/run-application/run-stateless-application-deployment/)、[コンテナの設定](/ja/docs/tasks/configure-pod-container/configure-pod-configmap/)、[リソースの管理](/ja/docs/concepts/cluster-administration/manage-deployment/)を参照してください。Ingressでは、Ingressコントローラーに依存しているいくつかのオプションの設定をするためにアノテーションを一般的に使用します。例としては、[rewrite-targetアノテーション](https://github.com/kubernetes/ingress-nginx/blob/master/docs/examples/rewrite/README.md)などがあります。[Ingressコントローラー](/ja/docs/concepts/services-networking/ingress-controllers)の種類が異なれば、サポートするアノテーションも異なります。サポートされているアノテーションについて学ぶためには、使用するIngressコントローラーのドキュメントを確認してください。

Reviewer: ```suggestion
Ingressには`apiVersion`、`kind`、`metadata`や`spec`フィールドが必要です。Ingressオブジェクトの名前は、有効な[DNSサブドメイン名](/ja/docs/concepts/overview/working-with-objects/names#dns-subdomain-names)である必要があります。設定ファイルに関する一般的な情報は、[アプリケーションのデプロイ](/ja/docs/tasks/run-application/run-stateless-application-deployment/)、[コンテナの設定](/ja/docs/tasks/configure-pod-container/configure-pod-configmap/)、[リソースの管理](/ja/docs/concepts/cluster-administration/manage-deployment/)を参照してください。Ingressでは、Ingressコントローラーに依存しているいくつかのオプションの設定をするためにアノテーションを一般的に使用します。例としては、[rewrite-targetアノテーション](https://github.com/kubernetes/ingress-nginx/blob/master/docs/examples/rewrite/README.md)などがあります。[Ingressコントローラー](/ja/docs/concepts/services-networking/ingress-controllers)の種類が異なれば、サポートするアノテーションも異なります。サポートされているアノテーションについて学ぶためには、使用するIngressコントローラーのドキュメントを確認してください。
```

Ingress [Spec](https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status)は、ロードバランサーやプロキシーサーバーを設定するために必要な全ての情報を持っています。最も重要なものとして、外部からくる全てのリクエストに対して一致したルールのリストを含みます。IngressリソースはHTTP(S)トラフィックに対してのルールのみサポートしています。

`ingressClassName`が省略された場合、[デフォルトのIngressClass](#default-ingress-class)を定義する必要があります。

> デフォルトの`IngressClass`を定義しなくても動作するIngressコントローラーがいくつかあります。例えば、Ingress-NGINXコントローラー は[フラグ](https://kubernetes.github.io/ingress-nginx/#what-is-the-flag-watch-ingress-without-class)

Reviewer 2: ```suggestion
デフォルトの`IngressClass`を定義しなくても動作するIngressコントローラーがいくつかあります。例えば、Ingress-NGINXコントローラーは[フラグ](https://kubernetes.github.io/ingress-nginx/#what-is-the-flag-watch-ingress-without-class)
```

`--watch-ingress-without-class`で設定できます。ただし、[下記](#default-ingress-class)のようにデフォルトの`IngressClass`を指定することを[推奨します](https://kubernetes.github.io/ingress-nginx/#i-have-only-one-instance-of-the-ingresss-nginx-controller-in-my-cluster-what-should-i-do)。

### Ingressのルール

各HTTPルールは以下の情報を含みます。

* オプションで設定可能なホスト名。上記のリソースの例では、ホスト名が指定されていないので、そのルールは指定されたIPアドレスを経由する全てのインバウンドHTTPトラフィックに適用されます。ホスト名が指定されていると(例: foo.bar.com)、そのルールは指定されたホストに対して適用されます。
* パスのリスト(例: `/testpath`)。各パスには`service.name`と`service.port.name`または`service.port.number`で定義されるバックエンドが関連づけられます。ロードバランサーがトラフィックを関連づけられたServiceに転送するために、外部からくるリクエストのホスト名とパスが条件と一致させる必要があります。
* バックエンドは[Serviceドキュメント](/ja/docs/concepts/services-networking/service/)に書かれているようなService名とポート名の組み合わせ、または{{< glossary_tooltip term_id="CustomResourceDefinition" text="CRD" >}}による[カスタムリソースバックエンド](#resource-backend)です。Ingressで設定されたホスト名とパスのルールに一致するHTTP(とHTTPS)のリクエストは、リスト内のバックエンドに対して送信されます。

Ingressコントローラーでは、`defaultBackend`が設定されていることがあります。これはSpec内で指定されているパスに一致しないようなリクエストのためのバックエンドです。

### デフォルトのバックエンド {#default-backend}

> ルールが設定されていないIngressは、全てのトラフィックをデフォルトのバックエンドに転送します。`.spec.defaultBackend`はその場合にリクエストを処理するバックエンドになります。`defaultBackend`は、[Ingressコントローラー](/ja/docs/concepts/services-networking/ingress-controllers)のオプション設定であり、Ingressリソースでは指定されていません。`.spec.rules`を設定しない場合、`.spec.defaultBackend`の設定は必須です。`defaultBackend`が設定されていない場合、どのルールにもマッチしないリクエストの処理は、Ingressコントローラに任されます(このケースをどう処理するかは、お使いのIngressコントローラのドキュメントを参照してください)。

Reviewer: ```suggestion
ルールが設定されていないIngressは、全てのトラフィックをデフォルトのバックエンドに転送します。`.spec.defaultBackend`はその場合にリクエストを処理するバックエンドになります。`defaultBackend`は、[Ingressコントローラー](/ja/docs/concepts/services-networking/ingress-controllers)のオプション設定であり、Ingressリソースでは指定されていません。`.spec.rules`を設定しない場合、`.spec.defaultBackend`の設定は必須です。`defaultBackend`が設定されていない場合、どのルールにもマッチしないリクエストの処理は、Ingressコントローラーに任されます(このケースをどう処理するかは、お使いのIngressコントローラーのドキュメントを参照してください)。
```

## Ingress Class

Ingressは異なったコントローラーで実装されうるため、しばしば異なった設定を必要とします。

> 各Ingressはクラス、つまりIngressClassリソースへの参照を指定する必要があります。IngressClassリソースには、このクラスを実装するコントローラの名前などの追加設定が含まれています。

Reviewer: ```suggestion
各Ingressはクラス、つまりIngressClassリソースへの参照を指定する必要があります。IngressClassリソースには、このクラスを実装するコントローラーの名前などの追加設定が含まれています。
```

パラメータの`kind`(および`apiGroup`)はクラスタースコープのAPI(カスタムリソースの場合もあり)を指し、パラメーターの`name`はそのAPIの特定のクラスタースコープのリソースを特定します。

例えば:
```yaml
---
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  name: external-lb-1
spec:
  controller: example.com/ingress-controller
  parameters:
    # このIngressClassのパラメーターは「external-config-1」という名前の
    # ClusterIngressParameter(APIグループk8s.example.net)で指定されています。この定義は、Kubernetesに
    # クラスタースコープのパラメーターリソースを探すように指示しています。
    scope: Cluster
    apiGroup: k8s.example.net
    kind: ClusterIngressParameter
    name: external-config-1
```
{{% /tab %}}
{{% tab name="Namespaceスコープ" %}}
{{< feature-state for_k8s_version="v1.23" state="stable" >}}

`.spec.parameters`フィールドを設定して`.spec.parameters.scope`フィールドを`Namespace`に設定した場合、IngressClassはNamespaceスコープのリソースを参照します。また`.spec.parameters`内の`namespace`フィールドには、使用するパラメーターが含まれているNamespaceを設定する必要があります。

パラメータの`kind`(および`apiGroup`)はNamespaceスコープのAPI(例えば:ConfigMap)を指し、パラメーターの`name`は`namespace`で指定したNamespace内の特定のリソースを特定します。

Namespaceスコープのパラメーターはクラスターオペレーターがワークロードに使用される設定(例えば:ロードバランサー設定、APIゲートウェイ定義)に対する制御を委譲するのに役立ちます。クラスタースコープパラメーターを使用した場合は以下のいずれかになります:

- クラスターオペレーターチームは、新しい設定変更が適用されるたびに、別のチームの変更内容を承認する必要があります。

> - クラスターオペレーターは、アプリケーションチームがクラスタースコープのパラメーターリソースに変更を加えることができるように、[RBAC](/ja/docs/reference/access-authn-authz/rbac/)RoleとRoleBindingといった、特定のアクセス制御を定義する必要があります。

Reviewer 3: ```suggestion
- クラスターオペレーターは、アプリケーションチームがクラスタースコープのパラメーターリソースに変更を加えることができるように、[RBAC](/ja/docs/reference/access-authn-authz/rbac/)のRoleやRoleBindingといった、特定のアクセス制御を定義する必要があります。
```
Proposal to be read more easier.

## Ingress Class

Ingressは異なったコントローラーで実装されうるため、しばしば異なった設定を必要とします。
各Ingressはクラス、つまりIngressClassリソースへの参照を指定する必要があります。IngressClassリソースには、このクラスを実装するコントローラの名前などの追加設定が含まれています。

{{< codenew file="service/networking/external-lb.yaml" >}}

IngressClassの`.spec.parameters`フィールドを使って、そのIngressClassに関連する設定を持っている別のリソースを参照することができます。

> 使用するパラメータの種類は、IngressClassの`.spec.controller`フィールドで指定したIngressコントローラに依存します。

Reviewer: ```suggestion
使用するパラメーターの種類は、IngressClassの`.spec.controller`フィールドで指定したIngressコントローラーに依存します。
```

### IngressClassスコープ

> ingressコントローラーによっては、クラスター全体で設定したパラメータを使用できる場合もあれば、1つの名前空間に対してのみ設定したパラメータを使用できる場合もあります。

Reviewer 2: ```suggestion
Ingressコントローラーによっては、クラスター全体で設定したパラメータを使用できる場合もあれば、1つの名前空間に対してのみ設定したパラメータを使用できる場合もあります。
```

> Ingressコントローラーによっては、クラスター全体で設定したパラメータを使用できる場合もあれば、1つの名前空間に対してのみ設定したパラメータを使用できる場合もあります。

Author: ```suggestion
Ingressコントローラーによっては、クラスター全体で設定したパラメータを使用できる場合もあれば、1つのNamespaceに対してのみ設定したパラメータを使用できる場合もあります。
```

> Ingressコントローラーによっては、クラスター全体で設定したパラメータを使用できる場合もあれば、1つのNamespaceに対してのみ設定したパラメータを使用できる場合もあります。

Reviewer: ```suggestion
Ingressコントローラーによっては、クラスター全体で設定したパラメーターを使用できる場合もあれば、1つのNamespaceに対してのみ設定したパラメーターを使用できる場合もあります。
```

{{< tabs name="tabs_ingressclass_parameter_scope" >}}
{{% tab name="クラスタースコープ" %}}

> IngressClassパラメータのデフォルトのスコープは、クラスター全体です。

Reviewer: ```suggestion
IngressClassパラメーターのデフォルトのスコープは、クラスター全体です。
```

`.spec.parameters`フィールドを設定して`.spec.parameters.scope`フィールドを設定しなかった場合、または`.spec.parameters.scope`を`Cluster`に設定した場合、IngressClassはクラスタースコープのリソースを参照します。

> パラメータの`kind`(および`apiGroup`)はクラスタースコープのAPI(カスタムリソースの場合もあり)を指し、パラメータの`name`はそのAPIの特定のクラスタスーコープのリソースを特定します。

Reviewer 2: ```suggestion
パラメータの`kind`(および`apiGroup`)はクラスタースコープのAPI(カスタムリソースの場合もあり)を指し、パラメータの`name`はそのAPIの特定のクラスタースコープのリソースを特定します。
```

> パラメータの`kind`(および`apiGroup`)はクラスタースコープのAPI(カスタムリソースの場合もあり)を指し、パラメータの`name`はそのAPIの特定のクラスタースコープのリソースを特定します。

Reviewer: ```suggestion
パラメータの`kind`(および`apiGroup`)はクラスタースコープのAPI(カスタムリソースの場合もあり)を指し、パラメーターの`name`はそのAPIの特定のクラスタースコープのリソースを特定します。
```

例えば:
```yaml
---
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  name: external-lb-1
spec:
  controller: example.com/ingress-controller
  parameters:

>     # このIngressClassのパラメータは「external-config-1」という名前の

Reviewer: ```suggestion
    # このIngressClassのパラメーターは「external-config-1」という名前の
```

    # ClusterIngressParameter(APIグループk8s.example.net)で指定されています。この定義は、Kubernetesに

>     # クラスタースコープのパラメータリソースを探すように指示しています。

Reviewer: ```suggestion
    # クラスタースコープのパラメーターリソースを探すように指示しています。
```

    scope: Cluster
    apiGroup: k8s.example.net
    kind: ClusterIngressParameter
    name: external-config-1
```
{{% /tab %}}

> {{% tab name="名前空間スコープ" %}}

Reviewer 2: In this case, I think it is easier to understand Namespace if it is translated as "Namespace" instead of "名前空間".
What do you think?

Author: Thank you for your kind instructions! That's certainly true. I'm with you. I'll go fix all of them.

> {{% tab name="名前空間スコープ" %}}

Author: ```suggestion
{{% tab name="Namespaceスコープ" %}}
```

{{< feature-state for_k8s_version="v1.23" state="stable" >}}

> `.spec.parameters`フィールドを設定して`.spec.parameters.scope`フィールドを`Namespace`に設定した場合、IngressClassは名前空間スコープのリソースを参照します。また`.spec.parameters`内の`namespace`フィールドには、使用するパラメータが含まれている名前空間を設定する必要があります。

Author: ```suggestion
`.spec.parameters`フィールドを設定して`.spec.parameters.scope`フィールドを`Namespace`に設定した場合、IngressClassはNamespaceスコープのリソースを参照します。また`.spec.parameters`内の`namespace`フィールドには、使用するパラメータが含まれているNamespaceを設定する必要があります。
```

> `.spec.parameters`フィールドを設定して`.spec.parameters.scope`フィールドを`Namespace`に設定した場合、IngressClassはNamespaceスコープのリソースを参照します。また`.spec.parameters`内の`namespace`フィールドには、使用するパラメータが含まれているNamespaceを設定する必要があります。

Reviewer: ```suggestion
`.spec.parameters`フィールドを設定して`.spec.parameters.scope`フィールドを`Namespace`に設定した場合、IngressClassはNamespaceスコープのリソースを参照します。また`.spec.parameters`内の`namespace`フィールドには、使用するパラメーターが含まれているNamespaceを設定する必要があります。
```

> パラメータの`kind`(および`apiGroup`)は名前空間スコープのAPI(例えば:ConfigMap)を指し、パラメータの`name`

Author: ```suggestion
パラメータの`kind`(および`apiGroup`)はNamespaceスコープのAPI(例えば:ConfigMap)を指し、パラメータの`name`
```

> は`namespace`で指定した名前空間内の特定のリソースを特定します。

Author: ```suggestion
は`namespace`で指定したNamespace内の特定のリソースを特定します。
```

> パラメータの`kind`(および`apiGroup`)はNamespaceスコープのAPI(例えば:ConfigMap)を指し、パラメータの`name`
> は`namespace`で指定したNamespace内の特定のリソースを特定します。

Reviewer: ```suggestion
パラメータの`kind`(および`apiGroup`)はNamespaceスコープのAPI(例えば:ConfigMap)を指し、パラメーターの`name`は`namespace`で指定したNamespace内の特定のリソースを特定します。
```

> 名前空間スコープのパラメータはクラスターオペレーターがワークロードに使用される設定(例えば:ロードバランサー設定、APIゲートウェイ定義)に対する制御を委譲するのに役立ちます。クラスタースコープパラメータを使用した場合は以下のいずれかになります:

Author: ```suggestion
Namespaceスコープのパラメータはクラスターオペレーターがワークロードに使用される設定(例えば:ロードバランサー設定、APIゲートウェイ定義)に対する制御を委譲するのに役立ちます。クラスタースコープパラメータを使用した場合は以下のいずれかになります:
```

> Namespaceスコープのパラメータはクラスターオペレーターがワークロードに使用される設定(例えば:ロードバランサー設定、APIゲートウェイ定義)に対する制御を委譲するのに役立ちます。クラスタースコープパラメータを使用した場合は以下のいずれかになります:

Reviewer: ```suggestion
Namespaceスコープのパラメーターはクラスターオペレーターがワークロードに使用される設定(例えば:ロードバランサー設定、APIゲートウェイ定義)に対する制御を委譲するのに役立ちます。クラスタースコープパラメーターを使用した場合は以下のいずれかになります:
```

- クラスターオペレーターチームは、新しい設定変更が適用されるたびに、別のチームの変更内容を承認する必要があります。

> - クラスターオペレーターは、アプリケーションチームがクラスタースコープのパラメータリソースに変更を加えることができるように、[RBAC](/ja/docs/reference/access-authn-authz/rbac/) RoleとRoleBindingといった、特定のアクセス制御を定義する必要があります。

Reviewer 2: ```suggestion
- クラスターオペレーターは、アプリケーションチームがクラスタースコープのパラメータリソースに変更を加えることができるように、[RBAC](/ja/docs/reference/access-authn-authz/rbac/)RoleとRoleBindingといった、特定のアクセス制御を定義する必要があります。
```

> - クラスターオペレーターは、アプリケーションチームがクラスタースコープのパラメータリソースに変更を加えることができるように、[RBAC](/ja/docs/reference/access-authn-authz/rbac/)RoleとRoleBindingといった、特定のアクセス制御を定義する必要があります。

Reviewer: ```suggestion
- クラスターオペレーターは、アプリケーションチームがクラスタースコープのパラメーターリソースに変更を加えることができるように、[RBAC](/ja/docs/reference/access-authn-authz/rbac/)RoleとRoleBindingといった、特定のアクセス制御を定義する必要があります。
```

IngressClass API自体は常にクラスタースコープです。

> 以下は名前空間スコープのパラメータを参照しているIngressClassの例です:

Author: ```suggestion
以下はNamespaceスコープのパラメータを参照しているIngressClassの例です:
```

> 以下はNamespaceスコープのパラメータを参照しているIngressClassの例です:

Reviewer: ```suggestion
以下はNamespaceスコープのパラメーターを参照しているIngressClassの例です:
```

```yaml
---
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  name: external-lb-2
spec:
  controller: example.com/ingress-controller
  parameters:

>     # このIngressClassのパラメータは「external-config」という名前の

Reviewer: ```suggestion
    # このIngressClassのパラメーターは「external-config」という名前の
```

    # IngressParameter(APIグループk8s.example.com)で指定されています。

>     # このリソースは「external-configuration」という名前空間にあります。

Author: ```suggestion
    # このリソースは「external-configuration」というNamespaceにあります。
```

{{< codenew file="service/networking/tls-example-ingress.yaml" >}}

{{< note >}}
サポートされるTLSの機能はIngressコントローラーによって違いがあります。利用する環境でTLSがどのように動作するかを理解するためには、[nginx](https://kubernetes.github.io/ingress-nginx/user-guide/tls/)や、[GCE](https://git.k8s.io/ingress-gce/README.md#frontend-https)、または他のプラットフォーム固有のIngressコントローラーのドキュメントを確認してください。
{{< /note >}}

### 負荷分散 {#load-balancing}

Ingressコントローラーは、負荷分散アルゴリズムやバックエンドの重みスキームなど、すべてのIngressに適用されるいくつかの負荷分散ポリシーの設定とともにブートストラップされます。発展した負荷分散のコンセプト(例: セッションの永続化、動的重み付けなど)はIngressによってサポートされていません。代わりに、それらの機能はService用のロードバランサーを介して利用できます。

> Ingressによってヘルスチェックの機能が直接に公開されていませんが、Kubernetesにおいて、同等の機能を提供する[Readiness Probe](/ja/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)のようなコンセプトが存在することは注目に値します。コントローラーがどのようにヘルスチェックを行うかについては、コントローラーのドキュメントを参照してください(例えば[nginx](https://git.k8s.io/ingress-nginx/README.md)、または[GCE](https://git.k8s.io/ingress-gce/README.md#health-checks))。

Reviewer: ```suggestion
ヘルスチェックの機能はIngressによって直接には公開されていませんが、Kubernetesにおいて、同等の機能を提供する[Readiness Probe](/ja/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)のようなコンセプトが存在することは注目に値します。コントローラーがどのようにヘルスチェックを行うかについては、コントローラーのドキュメントを参照してください(例えば[nginx](https://git.k8s.io/ingress-nginx/README.md)、または[GCE](https://git.k8s.io/ingress-gce/README.md#health-checks))。
```

## Ingressの代替案 {#alternatives}

> Ingressリソースを直接含まなくサービスを公開できる方法は複数あります。

Reviewer: ```suggestion
Ingressリソースを直接含まずにサービスを公開する方法は複数あります。
```
